### DISCUSSION TITLE
[BUG]: AtomWithStorage should not return initialvalue

### DISCUSSION THREAD

**Discussion by Pszz** [pmndrs/jotai#2622] [Bug report] [answered]
## Summary
```tsx
// When the user is not logged in, I set his sign to empty
const UserSign = atomWithStorage<string | null>('user_sign', null)

// hooks
const useLogin = () => { 
   const [sign, setSign] = useAtom(UserSign)
   console.log('sign::', sign)
   
   const login = async () => {
      await getUserToken()
      setSign('user sign token')
   }  
  
   useEffect(() => {
    // It always triggers the first time. 
    // In my opinion, when there is no content in storage, 
    // it has an initial value, and when there is content, the initial value should not be returned.

     if(sign === null) {
       login()
   }

}, [sign])
}
```

### Other

1. initialValue in atomWithStorage should not be required.
2. The third parameter is not required either.
3. I can't get around the second point when I only need to modify getOnInit:true in options.

**Comment by dai-shi** [comment]
Hi, thanks for reporting.
I'm not exactly sure if I understand the problem.
Can you elaborate the problem with `getOnInit: true`?

**Reply by Pszz** [reply]
I need to check whether the user is logged in when initializing the page, and if not, log in.
However, when there is an initial value, the user will be considered not logged in every time.

Currently, there are 3 variables in the `atomWithStorage` function.
`(key, initialValue, options)`

When I only need to configure `key` and `options`, I must fill in `initialValue`.

**Reply by dai-shi** [reply, marked as answer]
`initialValue` is required for the case there's no item in the storage.
I would do something like `('user_sign', undefined, { getOnInit: true })`.

**Reply by nivethan-me** [reply]
> `initialValue` is required for the case there's no item in the storage. I would do something like `('user_sign', undefined, { getOnInit: true })`.

what about storage? that's the 3rd parameter right?

if i try the following code 
```
export const accessTokenAtom = atomWithStorage(
  'accessToken',
  undefined,
  {getOnInit:true}
);
```
i'll get a typescript error saying

Object literal may only specify known properties, and 'getOnInit' does not exist in type 'SyncStorage<undefined>'

**Reply by dai-shi** [reply]
https://github.com/pmndrs/jotai/blob/a36209fd07fdc5526e3e35deaf38a1999cf0f6af/src/vanilla/utils/atomWithStorage.ts#L211-L216

Ah, it's the forth option.
